src/network/protocol/messages/getblocks.zig

```diff
@@ -0,0 +1,195 @@
+        }
+        try w.writeAll(&self.stop_hash);
+    }
+
+    /// Serialize a message as bytes and return them.
+    pub fn serialize(self: *const GetblocksMessage, allocator: std.mem.Allocator) ![]u8 {
+        const serialized_len = self.hintSerializedLen();
+
+        const ret = try allocator.alloc(u8, serialized_len);
+        errdefer allocator.free(ret);
+
+        try self.serializeToSlice(ret);
+
+        return ret;
+    }
+
+    pub fn deserializeReader(allocator: std.mem.Allocator, r: anytype) !GetblocksMessage {
+        comptime {
+            if (!std.meta.hasFn(@TypeOf(r), "readInt")) @compileError("Expects r to have fn 'readInt'.");
+            if (!std.meta.hasFn(@TypeOf(r), "readNoEof")) @compileError("Expects r to have fn 'readNoEof'.");
+        }
+
+        const buffer: []const u8 = try r.readAllAlloc(std.heap.page_allocator, MAX_SIZE + 1 );
+
+        // check message size
+        const msg_size = buffer.len;
+        if (msg_size > MAX_SIZE) {
+            return error.MessageTooLarge;
+        }
+
+        var gb: GetblocksMessage = undefined;
+        var fbs = std.io.fixedBufferStream(buffer);
+        var reader = fbs.reader();
+
+        gb.version = try reader.readInt(i32, .little);
+
+        // Read CompactSize hash_count
+        const compact_hash_count = try CompactSizeUint.decodeReader(reader);
+        gb.hash_count = compact_hash_count.value();
+
+        // Allocate space for header_hashes based on hash_count
+        const header_hashes = try allocator.alloc([32]u8, gb.hash_count);
+
+        for (header_hashes) |*hash| {
+            try reader.readNoEof(hash);
+        }
+        gb.header_hashes = header_hashes;
+
+        // Read the stop_hash (32 bytes)
+        try reader.readNoEof(&gb.stop_hash);
+        return gb;
+    }
+
+    /// Deserialize bytes into a `VersionMessage`
+    pub fn deserializeSlice(allocator: std.mem.Allocator, bytes: []const u8) !GetblocksMessage {
+        var fbs = std.io.fixedBufferStream(bytes);
+        const reader = fbs.reader();
+
+        return try GetblocksMessage.deserializeReader(allocator, reader);
+    }
+
+    pub fn hintSerializedLen(self: GetblocksMessage) usize {
+        const fixed_length = 4 + 32; // version (4 bytes) + stop_hash (32 bytes)
+        const compact_hash_count_len = CompactSizeUint.new(self.hash_count).hint_encoded_len();
+        const header_hashes_len = self.header_hashes.len * 32; // hash (32 bytes)
+        return fixed_length + compact_hash_count_len + header_hashes_len;
+    }
+
+    pub fn eql(self: *const GetblocksMessage, other: *const GetblocksMessage) bool {
+        if (self.version != other.version or self.hash_count != other.hash_count) {
+            return false;
+        }
+
+        if (self.header_hashes.len != other.header_hashes.len) {
+            return false;
+        }
+
+        for (self.header_hashes) |*hash| {
+            if (!std.mem.eql(u8, &hash.*, &other.header_hashes[0])) {
```

Comment:
`&other.header_hashes[0]`

You are comparing all of the self hashes to the first hashes in other. All the time

Reply:
yes true.